luavm/src/lib/lua_values/mod.rs

```diff
@@ -0,0 +1,754 @@
+        Ok(LuaVal::from(self.to_float()?.powf(other.to_float()?)))
+    }
+}
+
+impl From<i64> for LuaVal {
+    /// Create an integer LuaVal.
+    fn from(int: i64) -> Self {
+        LuaVal {
+            encoded_ptr: to_raw_ptr(int) ^ LuaValKind::INT as usize,
+            owns: true,
+        }
+    }
+}
+
+impl From<f64> for LuaVal {
+    /// Create an integer LuaVal.
+    fn from(float: f64) -> Self {
+        LuaVal {
+            encoded_ptr: to_raw_ptr(float) ^ LuaValKind::FLOAT as usize,
+            owns: true,
+        }
+    }
+}
+
+impl From<String> for LuaVal {
+    /// Create a string LuaVal.
+    fn from(string: String) -> Self {
+        LuaVal {
+            encoded_ptr: to_raw_ptr(string) ^ LuaValKind::STR as usize,
+            owns: true,
+        }
+    }
+}
+
+impl From<LuaTable> for LuaVal {
+    /// Create an integer LuaVal.
+    fn from(table: LuaTable) -> Self {
+        LuaVal {
+            encoded_ptr: to_raw_ptr(Gc::new(table)) ^ LuaValKind::TABLE as usize,
+            owns: true,
+        }
+    }
+}
+
+impl Drop for LuaVal {
+    fn drop(&mut self) {
+        if self.owns {
+            unsafe {
+                match self.kind() {
+                    // NIL is a nullptr, so there is no need to free
+                    LuaValKind::NIL => (),
+                    LuaValKind::INT => {
+                        Box::from_raw(i64_ptr(self.encoded_ptr));
+                    }
+                    LuaValKind::FLOAT => {
+                        Box::from_raw(f64_ptr(self.encoded_ptr));
+                    }
+                    LuaValKind::STR => {
+                        Box::from_raw(str_ptr(self.encoded_ptr));
+                    }
+                    LuaValKind::TABLE => {
+                        Box::from_raw(table_ptr(self.encoded_ptr));
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl Clone for LuaVal {
+    fn clone(&self) -> LuaVal {
+        let encoded_ptr = unsafe {
+            match self.kind() {
+                LuaValKind::NIL => 0,
+                LuaValKind::INT => to_raw_ptr(*i64_ptr(self.encoded_ptr)),
+                LuaValKind::FLOAT => to_raw_ptr(*f64_ptr(self.encoded_ptr)),
+                LuaValKind::STR => to_raw_ptr((*str_ptr(self.encoded_ptr)).clone()),
+                LuaValKind::TABLE => to_raw_ptr((*table_ptr(self.encoded_ptr)).clone()),
+            }
+        } ^ self.kind() as usize;
+        LuaVal {
+            encoded_ptr,
+            owns: true,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::{collections::HashMap, vec::Vec};
+
+    fn test_get_and_set_attr_errors(main: &mut LuaVal) {
+        assert_eq!(main.get_attr("foo").unwrap_err(), LuaError::GetAttrErr);
+        assert_eq!(
+            main.set_attr("foo", LuaVal::new()).unwrap_err(),
+            LuaError::SetAttrErr
+        );
+    }
+
+    fn test_shallow_copy_and_clone(main: &mut LuaVal) {
+        {
```

Comment:
Are the additional scopes needed here?

Reply:
It is kinda of an implicit test, where if the test crashes after this method is called, then it means the original `LuaVal` was not cloned/copied correctly.